### ISSUE TITLE
Display progress bars in taskbar/dock

### ISSUE THREAD

**Issue by Spekular** [LMMS/lmms#3768]
On Windows, applications can display a progress bar over the taskbar icon. For example, MPC-HC displays playback position when paused (left) and playing (right):
![Windows example](https://user-images.githubusercontent.com/8918426/29314118-db5e982e-81bc-11e7-8801-f0dd5c1475cc.png)

Ideally, progress bars in LMMS would also display on the taskbar. This would let users do something else while a project loads* or exports and still know when it's done.

This seems to be possible on macOS as well:
![macOS Example](https://sixcolors.com/images/content/stuck-progress-bars-6c.jpg)

Linux support seems more difficult, but Unity can have progress bars:
![Linux example](http://www.omgubuntu.co.uk/wp-content/uploads/2016/08/firefox-unity-progress-bar-1604.jpg)

*Right now project loading grabs window focus every time a task is completed, at least on Windows.

**Comment by Spekular** [comment]
FWIW I've found this, regarding Windows:

[TaskbarItemInfo.ProgressState Property (System.Windows.Shell)](https://msdn.microsoft.com/en-us/library/system.windows.shell.taskbariteminfo.progressstate(v=vs.110).aspx)

[TaskbarItemInfo.ProgressValue Property (System.Windows.Shell)](https://msdn.microsoft.com/en-us/library/system.windows.shell.taskbariteminfo.progressvalue(v=vs.110).aspx)

**Comment by tresf** [comment]
The macOS screenshot above was unusually large, I've shrunk it down a bit.

I like this idea.  Can you find anything on Gnome3?  [Unity is being dropped](https://betanews.com/2017/08/08/canonical-ubuntu-linux-gnome/)
, so it probably won't be worthwhile to write something against Unity specifically, but rather something that Gnome and KDE can benefit from (and hopefully others).

**Comment by Spekular** [comment]
As far as I can tell it's not available in gnome:

https://bugzilla.gnome.org/show_bug.cgi?id=701402

/Offtopic

I'm assuming you downloaded->cropped/resized->uploaded the macOS image to
resize it? The windows example is my own, so I made it suitably small, but
the other two I just searched up.

**Comment by tresf** [comment]
> I'm assuming you downloaded->cropped/resized->uploaded the macOS image to
resize it?

Yes.

> As far as I can tell it's not available in gnome:

Ok, let's use the original bug report to document this.
* Link the Windows API
* Try to find and link the macOS API
* Link the Ubuntu API with warning that it'll be removed in 17.04
* Link the Gnome bug report API

We can code the Windows, Mac and Ubuntu stuff and put the stubs in for Gnome.  From what I'm reading, KDE may already work with the Ubuntu implementation, so this would be worth testing on other desktops as well.

And as always, help/PRs are appreciated with this initiative.

**Comment by Spekular** [comment]
Ping to let email users know I've updated the first post.

**Comment by musikBear** [comment]
imo, not a needed feature 👎

**Comment by Spekular** [comment]
@musikBear is there any downside to it, though (other than opportunity
cost)? I'm sure there are people who would appreciate it, including myself
and tresf (afaik).

**Comment by tresf** [comment]
@musikBear this is a UI/UX feature and for long project exports (and perhaps large projects loading), it's valuable.   The purpose of the bug tracker is to document, code and test bugs.  Adding a thumbs down to a well thought proposal is OK, but please understand that it isn't very constructive.

**Comment by musikBear** [comment]
>  is there any downside to it, though (other than opportunity cost)?

Wouldent it prolong the loading, with yet another process / thread in that process?

**Comment by tresf** [comment]
> Wouldent it prolong the loading, with yet another process / thread in that process?

We already use progress indicators, this is just to display it in a progress bar as well.

**Comment by gi0e5b06** [comment]
http://doc.qt.io/qt-5/qsystemtrayicon.html
Don't use anything else.
(btw for Linux this a standard D-bus service)

**Comment by tresf** [comment]
> http://doc.qt.io/qt-5/qsystemtrayicon.html
> Don't use anything else.

@gi0e5b06 thanks for the link, but can you please explain how this is related to the dock area?  The system tray is a completely separate part of the desktop and also one that we do not leverage (not have we had a need to leverage).

**Comment by gi0e5b06** [comment]
@tresf Displaying progress in the taskbar is neither common nor standard. Most apps use notifications instead (often displayed in the system tray). My point was to suggest not to use anything which is not Qt.

**Comment by tresf** [comment]
> Displaying progress in the taskbar is neither common nor standard.

The code may not (yet) be standard, but the UX is and we're talking about UX improvements.  Gnome doesn't recognize it, but [hopefully Ubuntu changes that as they switch 17.04 to gnome/wayland](https://wiki.ubuntu.com/ArtfulAardvark/ReleaseNotes).  Gnome also doesn't believe in minimize buttons, so they're the ridiculous exception, not the standard.  I digress.  We can code this and it's not a lot of work.

Examples of standard UX that thinks the launcher deserves the progress indicator... 
* iOS devices do this when apps are installing.  iOS apps represent [about 20% of the phone market.  About 10% of devices](https://en.wikipedia.org/wiki/Usage_share_of_operating_systems#Desktop_and_laptop_computers).
* Browsers do this when files are downloading on Windows and MacOS, the taskbar progress is used for Chrome, IE, Safari and Firefox.
* Desktops do this when large files are being moved or copied, the taskbar progress is used to display overall copy/move/io progress.

In regards to the APIs that are available... [Win32](http://doc.qt.io/qt-5/qwintaskbarprogress.html), [Cocoa (MacOS)](https://stackoverflow.com/questions/43347722/qt-show-progress-bar-in-dock-macos/43372236) and [Unity (Ubuntu)](https://wiki.ubuntu.com/Unity/LauncherAPI) all support this, just not in a standard API.

In summary, I strongly disagree with the System Tray placement.  I feel this is non-standard UX for an interactive desktop application and feel LMMS NOT belong in the tray.  Please provide a better use-case for your argument.  I've provided (3) three standard and common UX examples where the task bar is the right place for status updates on lengthy operations.

On a side-note, I'm the author of [a program](https://github.com/qzind/tray) **that runs exclusively in the system tray**, so I feel I'm a good judge of when the tray should as well as when the tray should not be used in a user environment.

**Comment by NickAcPT** [comment]
I think this is a good idea, **but not for LMMS**.
In order to make this proposal into reality, the developers would need to depend on **different OS specific code**.
We, people, need to think ahead.
What if some people grouped up to port LMMS to Android/iOS devices? What if they wanted to have one source code that would produce binaries to Windows/Linux/MacOSX/Android/iOS, all in one codebase? That would be hard to maintain. And I'm not talking about now. I'm talking about the future.

###### I figured I would give my two cents about this. Sorry if my english is broken. 😄

**Comment by gi0e5b06** [comment]
@tresf That was a long answer :)
I was suggesting to notify the same way as when you copy files.
Anyway, what you're looking at is this:
https://code.qt.io/cgit/qt-creator/qt-creator.git/tree/src/plugins/coreplugin/progressmanager

**Comment by tresf** [comment]
> What if some people grouped up to port LMMS to Android/iOS devices? What if they wanted to have one source code that would produce binaries to Windows/Linux/MacOSX/Android/iOS, all in one codebase? 

We're [bikeshedding](https://en.wiktionary.org/wiki/bikeshedding).  Again. The obstacles to Android and iOS are far greater than a stupid progress indicator.  Sound-backend alone is already largely platform dependent as is MIDI support.  Although we appreciate articulated answer, please stop speculating, it takes away from the otherwise well thought out conversation.

@gi0e5b06 thanks for the code. 👍 

> I was suggesting to notify the same way as when you copy files.

I'm not sure I read that from your posts, but thanks for the nudge in the right direction.

**Comment by NickAcPT** [comment]
> The obstacles to Android and iOS are far greater than a stupid progress indicator. Sound-backend alone is already largely platform dependent as is MIDI support (better yet the pile of libraries we compile into our code).

Android Marshmallow introduced a MIDI API.
I still haven't searched if there is a MIDI API for iOS devices

> Did you know Android is Java and iOS is Objective-C?

Yeah, I know that.
When I talked about "Android and iOS with the same codebase", I was subconsciously thinking about Xamarin. In this case, the downside of Xamarin, is that you'll need to  migrate code to c#.
Also, with a small search I got to [this](https://stackoverflow.com/questions/18849620/xamarin-and-c-code) question on StackOverflow that talks about using C++ code on Xamarin.

###### Sorry for bothering you and thank you for your fast reply

**Comment by eagles051387** [comment]
Your iOS statement is inaccurate. In the early days iOS was objective c only now you can develop with c/c++ as well for iOS

**Comment by softrabbit** [comment]
> When I talked about "Android and iOS with the same codebase", I was subconsciously thinking about Xamarin. In this case, the downside of Xamarin, is that you'll need to migrate code to c#.
Also, with a small search I got to this question on StackOverflow that talks about using C++ code on Xamarin.

Qt should cover those two mobile OSes, see http://doc.qt.io/qt-5/supported-platforms.html

**Comment by tresf** [comment]
> Qt should cover those two mobile OSes, see http://doc.qt.io/qt-5/supported-platforms.html

👍 

> Your iOS statement is inaccurate. In the early days iOS was objective c only now you can develop with c/c++ as well for iOS

Thanks for joining the discussion. :)  Thanks for the information, that's useful if we ever port to mobile.  Let's not minimize the effort though, please.  This run-on conversation started when it was suggest we do NOT use platform specific code for a particular feature due to the impact it would have on other OSs, specifically mobile OSs.

Although we'd garner a bunch of happy users if we ported to Android, the CPU architecture tends to be ARM and we would need to do a lot of legwork, such as porting our dependent libraries over as well.  Also, we rely on some disk IO features that would break in a sandbox.  The point here is, the progress indicator is 1. Standard.  2. Not a lot of work.

Cherry-picking argument is great for education, but if someone were to look back at this thread a year from now, it's all just offtopic banter. <3

**Comment by follower** [comment]
FWIW if Qt Creator's progress manager linked above isn't suitable, then (mis-)use of the following (on Mac OS X) might be another option:

 * `QWindow::setWindowIcon()` [[via](http://doc.qt.io/qt-5/osx-issues.html#dock)]

 * `QtMac::setBadgeLabelText()` [[via](http://doc.qt.io/qt-5/qtmac.html#setBadgeLabelText)]

**Comment by russiankumar** [comment]
Anyone assigned/working on this? Was looking to try and implement this if it is still a wanted feature.

Regarding the porting to mobile platforms mentioned here, I believe a lot of the UI will have to be rewritten anyways, so I suppose one progress bar isn't really a big issue in that respect.

**Comment by Spekular** [comment]
No one is working on it as far as I'm aware, and yes I still think it's a good idea :)
